### PR TITLE
Support not Insertable/Updateable columns for entities with `JOINED` inheritance type

### PR DESCRIPTION
**Pull request by e-vil-dev**
Fix for https://github.com/doctrine/orm/issues/9467 when insert and update entity with `joined` inheritance + update values from database

1. Entities with `#[InheritanceType('JOINED')]` attribute process by `JoinedSubclassPersister`, according to `UnitOfWork::getEntityPersister()`:
    https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/UnitOfWork.php#L3227-L3228
1. The persister override `getInsertColumnList()` without excluding `notInsertable` columns like parent do:
    https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php#L1478
  so when it use prepared sql-statement with parameters prepared by parent (`AbstractEntityInheritancePersister::prepareInsertData()` reuse `BasicEntityPersister::prepareInsertData()` reuse `BasicEntityPersister::prepareUpdateData()` with `$isInsert` mode):
    https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php#L640-L659
  we get error from database, e.g. postgresql:
    ```
    DOException: SQLSTATE[08P01]: <<Unknown error>>: 7 ERROR:  bind message supplies 4 parameters, but prepared statement "" requires 6
    ```
   So fixed `JoinedSubclassPersister::getInsertColumnList()`. There is no such error for update queries, because parent methods used in `JoinedSubclassPersister::update()`: `BasicEntityPersister::prepareUpdateData()` and `BasicEntityPersister::updateTable()`.
1. There were problems with logic to update entity properties with database values in generated columns (`notInsertable`|`notUpdatable`|`version`) after flush - `JoinedSubclassPersister::assignDefaultVersionAndUpsertableValues()`:
    https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php#L548-L555
    - flag which activates it (`ClassMetadataInfo::$requiresFetchAfterChange`) in both cases ([executeInserts()](https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php#L170-L171) and [update()](https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php#L224-L235)) doesnt inherited from entity parent-class. This fixed by tracking of parent mappings in `ClassMetadataInfo::addInheritedFieldMapping()` like it done for child entity in `mapField()` (may be add setter for [factory](https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L105-L108)):
    https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L2814-L2820
    - it use parent select query to get generated columns, while it could be in both tables (child and parent), so override query builder in persister - `JoinedSubclassPersister::fetchVersionAndNotUpsertableValues()` - where added join (reuse `JoinedSubclassPersister::getJoinSql()`) and column aliases. Make parent `extractIdentifierTypes()` and `$identifierFlattener` make them protected.
    - because generated column may be in child tables logic should be after insert into child tables in persister `insert()` method to avoid `LengthException('Unexpected empty result for database query.')`.
1. For functional tests added `JoinedInheritanceRoot` entity, with children for each case and `JoinedInheritanceChild` to test root entity columns. Declare database default column values to avoid platform-specific `columnDefinition`. Use `generated: 'ALWAYS'` in hope to obtain error from `update()` method

### CODE REVIEW

**Comment by mrVrAlex** [comment]
P.S. This is replacement for PR https://github.com/doctrine/orm/pull/9486 but with functional tests and covered more problem of combination cases (like nonInsertable column in parent entity) which not worked even after apply fixes from previous PR.

**Comment by mrVrAlex** [comment]
@greg0ire Can you approve workflow runs checks on this? 
Also maybe need choose 2.15.x as base branch for this, or patch release it is ok?

**Comment by greg0ire** [comment]
Based on the title, it's not really a fix, but a new feature, so I'd say 2.15.x

**Comment by mrVrAlex** [comment]
@greg0ire Done. Please approve CI again (also failed tests for php7.2 was fixed).
I think it's ready for review now)

**Comment by greg0ire** [comment]
@mehldau please review. Also please advise on whether this is a patch or an improvement. When reading the description, I'm now on the verge of considering this as a fix, since it's a special case of something supposed to work, and not something extra you did not do deliberately.

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:186]:
```suggestion
    final protected $identifierFlattener;
```

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]:
```suggestion
    final protected function extractIdentifierTypes(array $id, ClassMetadata $versionedClass): array
```

**Comment by mrVrAlex** [comment]
@greg0ire As I understand we wait @mehldau decision? 
Just until we wait - 2.15 was already released)

**Comment by greg0ire** [comment]
@mehldau doesn't seem to respond, so let's not wait for them.

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:186]: Why was this marked as resolved?

File: tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php
Comment (greg0ire) [inline tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php]: You can use linebreaks here, can't you?

File: tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceNonUpdatableColumn.php
Comment (greg0ire) [inline tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceNonUpdatableColumn.php]: Can you please read and apply https://github.com/doctrine/orm/pull/10683#issuecomment-1537466852 ?

File: tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php
Comment (mrVrAlex) [reply tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php]: Then all PHP file will be not valid for < PHP 8.0 (multiline attribute), it's ok?)

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php
Comment (mrVrAlex) [reply lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:186]: Because of:

> PHP Fatal error:  Cannot declare property Doctrine\ORM\Persisters\Entity\BasicEntityPersister::$identifierFlattener final, the final modifier is allowed only for methods, classes, and class constants in /home/runner/work/orm/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php on line 186

File: tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php
Comment (greg0ire) [reply tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceRoot.php]: Oh right, please ignore my comment then.

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php
Comment (greg0ire) [reply lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:186]: Oh right… don't know where I got that from!

File: tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceNonUpdatableColumn.php
Comment (mrVrAlex) [reply tests/Doctrine/Tests/Models/Upsertable/JoinedInheritanceNonUpdatableColumn.php]: Done.

**Comment by greg0ire** [comment]
Please address the coding standard issues.

**Comment by greg0ire** [comment]
> Please address the coding standard issues.

I think they are not caused by you, I'll handle them separately.

**Comment by greg0ire** [comment]
Fixed. You just need to rebase.

**Comment by mrVrAlex** [comment]
> Fixed. You just need to rebase.

Rebased.

File: lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php]:
```suggestion
     * {@inheritDoc}
```

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php
Comment (SenseException) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php]:
```suggestion
        $this->_em->clear();
        $entity = $this->_em->find(JoinedInheritanceChild::class, $entity->id);
        // check fetch generated values override prefilled for notUpdatable
```

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php
Comment (SenseException) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php]: "Check for no changes for ..."?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php
Comment (SenseException) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php]: But `rootField` should have changed, right?

**Review by SenseException — CHANGES_REQUESTED** [review]
Sometimes you use `clear` and `find` in the tests and sometimes you don't. Could you please use them all the same in the test methods?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php
Comment (e-vil-dev) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH9467/GH9467Test.php]: Yes. I'll add this check, just tried to keep the test methods short

**Comment by e-vil-dev** [comment]
> Sometimes you use `clear` and `find` in the tests and sometimes you don't. Could you please use them all the same in the test methods?

Ok. Testing before `clear` checks that `flush` may fetch not Insertable/Updatable properties, and after `clear` + `find` that these values are indeed from the database. Also to test update query results, I skip `clear` to avoid `find` by reuse just inserted entity.
I'll try improve readability of unit-test

**Comment by mrVrAlex** [comment]
So guys, what's next step on this? 
BTW: We use these fixes on production (big) project already more then 2 months without problem.
And maybe after merge it we will report (and fix) the next issue with unnecessary `UPDATE xx SET id = yyy WHERE id = yyy` queries in some such cases (with joined inheritance).

**Review by mrVrAlex — APPROVED** [review]
@SenseException @derrabus @greg0ire 
Maybe already need rebase this to 2.16.x branch? What you think (how long you will hold it)?

**Comment by greg0ire** [comment]
If it is not a bugfix then yes, it should be rebased on 2.16.x. I don't understand your second question.

**Comment by mrVrAlex** [comment]
I'm not sure what it is: bugfix or improvement (as you mention it 2 month ago about it).
I know one thing: generated columns not works (since this feature was appear in [v2.11](https://github.com/doctrine/orm/pull/9118)) with JOINED inheritance type.
We ready to move it to any release. But IMHO: developers expect (https://github.com/doctrine/orm/issues/9467) what virtual / generated columns should work with any exist features (entity type) in doctrine. When we test fixes in previous PR (https://github.com/doctrine/orm/pull/9486) we discovered what not all cases covered by it, so we re-writed replacement solution (and it worked on our projects). 
Now just need hard maintainer decision where it should merge

**Comment by greg0ire** [comment]
@SenseException please give this another review.

**Comment by mrVrAlex** [comment]
@SenseException ping)

**Comment by mrVrAlex** [comment]
@SenseException @greg0ire so what you decided with it, remain 2.15.x branch (release in next v2.15.4?) or should rebase to 2.16 branch?
Let's go finally somewhere and will continue thinking further)

**Comment by SenseException** [comment]
> Based on the title, it's not really a fix

@greg0ire @mrVrAlex The description in #9467 sounds more like a feature so, without taking a deeper look, I'd say rebase to 2.16. Otherwise this looks good to me.

**Comment by e-vil-dev** [comment]
@SenseException rebased to 2.16

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks everyone!

**Comment by rcadhikari** [comment]
I had similar issue, however setting the `insertable` & `updatable` to `false` at `ORM\Column` level helped me to resolve the issue.

for example.
```
/**
     * My Column desc
     *
     * @var integer|null
     * @ORM\Column(
     *     name="your_column_name",
     *     type="integer",
     *     length=12,
     *     nullable=true,
     *     insertable=false,
     *     updatable=false,
     * )
     */
    protected $yourColumnName;
```

**Pull request merged**
